### PR TITLE
Certificate Refactor: Certificate Response, Renewals, Async Support

### PR DESCRIPTION
**Pull request by cristiangraz**
This is a PR for the conversation started in issue #11. Unit tests are in progress pending review from @ericchiang.

Details:
- `NewCertificate` now returns `CertificateResponse` struct
- `NewCertificate` now supports asynchronous certificates (Not yet in Boulder, but based on spec)
- If `NewCertificate` call results in a synchronous certificate, `CertificateResponse.Certificate` is populated with `*x509.Certificate`. `URI` and `StableURI` are both set. The issuer url is set based on the rel="up" Link header. (see below for more details
- If `NewCertificate` call results in an empty response body, `CertificateResponse` is populated with `RetryAfter` and `URI`.
- `RenewCertificate` method
- `RevokeCertificate` method

Methods on `CertificateResponse`:
- `CertificateResponse` contains three methods: `Retry`, `RetryPoll`, and `Bundle`
- `Retry` will handle implementation details to retry the call to `CertificateResponse.URI`. User triggers it based on number of seconds passed in `CertificateResponse.RetryAfter`. Behavior here is the same as `NewCertificate` depending on if the certificate is available or requires another try.
- `RetryPoll` will poll retries `maxRetries` times
- `Bundle` allows for bundling the certificate with the cross-signed Issuer. This is a convenience method and puts the burden on the user to call, but the implementation details are covered in the library.

### CODE REVIEW

**Comment by cristiangraz** [comment]
@ericchiang Any thoughts?

**Comment by ericchiang** [comment]
@cristiangraz sorry have been busy. Will look this over today :)

File: README.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ import (
     "github.com/ericchiang/letsencrypt"
 )
 
-var supportedChallengs = []string{
+var supportedChallenges = []string{
```
Comment (ericchiang) [inline README.md:30]: can you make things like spelling corrections their own pull requests?

File: certificate.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+func (c *CertificateResponse) Retry() error {
+	if c.IsAvailable() {
+		return errors.New("Aborting retry request. Certificate is already available")
+	}
+
+	if c.URI == "" {
+		return errors.New("Could not make retry request. No URI available")
+	}
+
+	resp, err := http.Get(c.URI)
```
Comment (ericchiang) [inline certificate.go:66]: the actual client holds an `http.Client` which makes all certificate requests. Would making `Bundle()` and `Retry()` functions on the client make more sense so it can reuse that client?

File: certificate.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	return fmt.Errorf("Retry expected status code of %d or %d, given %d", http.StatusOK, http.StatusAccepted, resp.StatusCode)
+}
+
+// RetryPoll will attempt to retrieve asynchronous certificates maxRetries times
+// and sleep for RetryAfter seconds between requests.
+// This method will not make the initial request until the initial
+// RetryAfter period elapses.
+func (c *CertificateResponse) RetryPoll(maxRetries int) error {
```
Comment (ericchiang) [inline certificate.go:117]: Have been leaning against adding methods like this. It's extremely sensitive to configuration: how long to wait, how often to retry, how to cancel, etc. Since it's so little code I'd rather have package users implement this themselves.

**Comment by ericchiang** [comment]
Overall looks really good!

General comment. Would it be possible for you to clean up your git commits? You've added stuff like [this](https://github.com/blacklightcms/letsencrypt/commit/cd01a76df4b9c7317832a856353a89368ab42725).

Since this is really only one change, it seems reasonable to roll it into a single commit.

File: certificate.go (hunk 2 above)
Comment (cristiangraz) [inline certificate.go:66]: I decided to make these on the CertificateResponse struct because that's the struct that has all the necessary info and retrying/bundling the response made more sense to me then retrying/bundling against the client. I'm happy to put in the client if you'd prefer that route -- passing in the CertificateResponse as a parameter would probably help with the readability of what you are retrying/bundling. I'll update.

File: certificate.go (hunk 3 above)
Comment (cristiangraz) [inline certificate.go:117]: I agree about keeping out certain methods that are easily implemented by the user, but I think the library should be consistent. The ChallengeReady method does exactly this -- but without any outside options from the user. That one is harder to implement than RetryPoll since RetryPoll lets you check IsAvailable to see if a certificate was returned, but I think this is confusing. I'm happy to remove RetryPoll as it's very easy to implement for the user, but something to consider.

**Comment by cristiangraz** [comment]
Thanks for reviewing! I'll make the changes, add unit tests, and put this all into a single commit to clean things up. I needed to get something into production and this branch ended up being a gradual progression. Let me know if you have an easier solution to clean things up, otherwise I'll create a new PR to supersede this one.

**Comment by cristiangraz** [comment]
This one is superseded by #14
